Case-insensitive identifiers

yariks5s opened the pull request:
### Changelog category (leave one):
- Improvement


### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Added setting `case_insensitive_names = 'standard'` for SQL-standard-like case-insensitive identifier resolution in SELECT analysis. In this mode, unquoted and backtick-quoted database, table, column, alias, and CTE references are matched case-insensitively while double-quoted identifiers remain exact; exact-case matches win and remaining case-only collisions are reported as ambiguous. INSERT targets and DDL statements continue to use exact catalog lookups. Closes https://github.com/ClickHouse/ClickHouse/issues/33935

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

yariks5s committed:
Case-insensitive identifiers

yariks5s committed:
small fixups

yariks5s committed:
Small fixes

yariks5s committed:
Fast tests fixups

yariks5s committed:
Tests + tidy

yariks5s committed:
Fix test errorcode

yariks5s committed:
replace a few more errorcodes

yariks5s committed:
Clearer logic

yariks5s committed:
Make the code simple

yariks5s committed:
Fix the bug about incorrect resolving

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/case-insensitive-identifiers

yariks5s committed:
Fix some bugs + move settings to 26.1

yariks5s committed:
fix tidy

yariks5s committed:
Correct fallback conditions

yariks5s committed:
Remove bad comment

jkartseva commented on src/Analyzer/Resolve/IdentifierResolver.cpp:
```diff
@@ -146,6 +152,149 @@ QueryTreeNodePtr IdentifierResolver::wrapExpressionNodeInTupleElement(QueryTreeN
+
+    bool operator<(const QualifiedTable & other) const
+    {
+        return database < other.database || (database == other.database && table < other.table);
```
nit: can be `std::tie(database, table) < std::tie(other.database, other.table)`

jkartseva commented on src/Analyzer/Resolve/IdentifierResolver.cpp:
```diff
@@ -613,22 +782,67 @@ IdentifierResolveResult IdentifierResolver::tryResolveIdentifierFromStorage(
+        else if (ci_col_name_matches.size() > 1)
+        {
+            throw Exception(
+                ErrorCodes::AMBIGUOUS_IDENTIFIER,
```
There is also the `AMBIGUOUS_COLUMN_NAME` error, but probably `AMBIGUOUS_IDENTIFIER` is better for consistency with database and table errors.

jkartseva reviewed (approved):
LGTM

yariks5s committed:
Addressing review

yariks5s committed:
Merge branch 'master' into yarik/case-insensitive-identifiers

yariks5s committed:
Fix build

yariks5s committed:
Merge branch 'yarik/case-insensitive-identifiers' of https://github.com/ClickHouse/ClickHouse into yarik/case-insensitive-identifiers

yariks5s committed:
Merge branch 'master' into yarik/case-insensitive-identifiers

novikd commented on src/Interpreters/DatabaseCatalog.cpp:
```diff
@@ -842,6 +843,20 @@ Databases DatabaseCatalog::getDatabases(GetDatabasesOptions options) const
+    {
+        if (Poco::toLower(name) == target)
+            result.push_back(name);
+    }
```
But maybe it should have another type to add a check if there are duplicates

novikd commented on src/Analyzer/Resolve/IdentifierResolver.cpp:
```diff
@@ -613,22 +782,67 @@ IdentifierResolveResult IdentifierResolver::tryResolveIdentifierFromStorage(
-            table_expression_node,
-            table_expression_data,
-            valid_identifiers);
+        /// case-insensitive fallback for columns
```
Subcolumn lookup in the top is not handled.

novikd commented on src/Analyzer/Resolve/IdentifierResolver.cpp:
```diff
@@ -733,7 +947,39 @@ IdentifierResolveResult IdentifierResolver::tryResolveIdentifierFromTableExpress
     }
 
     if (identifier.getPartsSize() == 1)
```
Cases in the bottom use table and database names and they are not handled.

novikd replied:
Following function in the top would break:
* `table_expression_data.hasFullIdentifierName(IdentifierView(identifier))`
* `table_expression_data.canBindIdentifier(IdentifierView(identifier))`

novikd reviewed (commented):
PR requires changes. Please, consider:
1. Filling table expression data in case-insensitive way in QueryAnalyzer.
2. Maybe aliases also should work in case-insensitive pashion.
3. Other databases may use upper-case -- may be we should as well.

yariks5s committed:
Revised logic

yariks5s committed:
Merge branch 'yarik/case-insensitive-identifiers' of https://github.com/ClickHouse/ClickHouse into yarik/case-insensitive-identifiers

yariks5s committed:
Merge branch 'master' into yarik/case-insensitive-identifiers

yariks5s committed:
Tests.

yariks5s committed:
Merge branch 'yarik/case-insensitive-identifiers' of https://github.com/ClickHouse/ClickHouse into yarik/case-insensitive-identifiers

yariks5s committed:
Minor

yariks5s committed:
Whitespaces

yariks5s committed:
Fix typo

yariks5s committed:
Self-review

yariks5s committed:
Style

yariks5s committed:
Review pt. 2

yariks5s committed:
Fix build

yariks5s committed:
Review pt. 3

yariks5s committed:
Review pt. 3 (case-insensitive db and table names)

yariks5s committed:
Build fix + tests

yariks5s committed:
Self-review pt. 5

yariks5s committed:
Style

yariks5s committed:
Do not hash quotations

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/case-insensitive-identifiers

yariks5s committed:
Minor (pt. 6)

yariks5s committed:
Merge branch 'master' into yarik/case-insensitive-identifiers

alexey-milovidov committed:
Merge origin/master into yarik/case-insensitive-identifiers

Resolved conflicts:
- `IdentifierNode.cpp`: adapted `setQuoteStyles` to use `make_intrusive` instead of `std::make_shared`
- `QueryAnalyzer.cpp`: combined PR's `is_part_double_quoted` tracking with master's `allow_to_resolve_niladic_functions` parameter; added both new extern settings
- `Settings.h`: added both `CaseInsensitiveNames` and `DeduplicateInsertMode` setting types
- `ExpressionElementParsers.cpp`: adapted `setQuoteStyle`/`setQuoteStyles` calls to use `make_intrusive`

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/case-insensitive-identifiers

# Conflicts:
#	src/Analyzer/Resolve/QueryAnalyzer.cpp

yariks5s committed:
Update SettingsChangesHistory.cpp

yariks5s committed:
Address review: per-part quote checks and expression-arg registration

- `tryBindIdentifierToTableExpression` /
  `tryResolveIdentifierFromTableExpression`: distinguish per-part quoting
  for qualifier matches (db / table / alias) via `isPartDoubleQuoted(i)`
  and keep `isLastPartDoubleQuoted` for the column-name lookup. Fixes
  mixed-quoting like `SELECT T."Col" FROM test_compound AS t` where the
  unquoted `T` qualifier must remain case-insensitive even though `"Col"`
  is double-quoted.

- `resolveFunction.cpp` joinGet path: derive
  `database_name_case_insensitive` / `table_name_case_insensitive` from
  `IdentifierNode::isPartDoubleQuoted(0/1)` instead of applying
  `standard_mode` to both parts unconditionally.

- `resolveInterpolateColumnsNodeList` and recursive-CTE setup: route
  expression-argument inserts through `addExpressionArgument` so the
  lowercase index used by case-insensitive `findExpressionArgument` stays
  consistent.

See review comments:
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3274219462
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3274220175
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3274458609

yariks5s committed:
Merge branch 'yarik/case-insensitive-identifiers' of github.com:ClickHouse/ClickHouse into yarik/case-insensitive-identifiers

yariks5s committed:
Address review: keep lowercase indexes in sync, fix duplicate-alias false ambiguity

- `QueryExpressionsAliasVisitor`: only call `registerAliasCaseInsensitive`
  when the corresponding primary `emplace` succeeded. Previously a query
  like `SELECT 1 AS x, 1 AS x, x` left the lowercase index with
  `x -> ["x", "x"]` and `findCaseInsensitive` reported it as ambiguous.

- `TableExpressionsAliasVisitor` and `add_table_expression_alias_into_scope`
  in `QueryAnalyzer::resolveQuery`: also write the alias into
  `lowercase_table_alias_to_originals` in standard mode. Without this,
  unquoted table-alias resolution missed e.g.
  `SELECT * FROM test_table AS t1, t1`.

- `QueryAnalyzer::resolveQuery`: the transitive-aliases swap-out/swap-in
  around the join-tree visit now also saves and restores
  `lowercase_table_alias_to_originals` so newly-registered lowercase
  entries do not leak past the swap-back.

- `DatabaseCatalog::updateDatabaseName`: maintain
  `lowercase_db_to_original_names` on `RENAME DATABASE`, otherwise
  case-insensitive lookups still pointed to the old database name.

See review comments:
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3275837378
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3275837876
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3275838286

yariks5s committed:
Preserve per-part quote styles in ASTIdentifier::createTable

The `joinGet` / `dictGet` / `IN` argument normalization (see
`MarkTableIdentifiersVisitor`) replaces the original `ASTIdentifier` with
an `ASTTableIdentifier` via `ASTIdentifier::createTable`. The previous
implementation only copied `name_parts`, dropping `quote_styles` on the
floor.

Without quote styles, downstream resolution loses the information needed
to honor per-part case sensitivity. For example,
`joinGet("Db"."Table", ...)` in `standard` mode reached the analyzer with
both parts looking unquoted and was therefore resolved
case-insensitively, even though both parts were explicitly double-quoted.

Copy `quote_styles` from `ASTIdentifier` onto the produced
`ASTTableIdentifier` so the analyzer's per-part `isPartDoubleQuoted`
checks see the original quoting.

See review comment:
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3276095428

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/case-insensitive-identifiers

# Conflicts:
#	src/Interpreters/DatabaseCatalog.h
#	src/Parsers/ExpressionElementParsers.cpp

yariks5s committed:
Address review: quote-style aware CTE/alias resolution and qualified COLUMNS

- Alias matching in tryResolveIdentifierFromAliases now uses isPartDoubleQuoted(0)
  for the alias part instead of isLastPartDoubleQuoted, so T."Col" still matches
  unquoted alias t in standard mode
- ParserQualifiedColumnsMatcher copies quote_styles when rebuilding the qualifier
  identifier (after dropping the trailing COLUMNS part), so "T".COLUMNS(...) stays
  case-sensitive in standard mode
- joinGet/dictGet initial expression lookup propagates per-part quote styles into
  IdentifierLookup so quoted args do not match unquoted aliases case-insensitively
- CTE registration keeps original case as the primary key and tracks lowercase
  variants in a parallel index; tryResolveIdentifierFromCTE consults the lowercase
  index only for unquoted lookups, so WITH "MyCte" AS ... SELECT * FROM "MyCte"
  resolves and "MyCte" / "mycte" no longer collide
- ASTWithAlias + IQueryTreeNode track whether the alias was double-quoted, and
  QueryExpressionsAliasVisitor skips the lowercase index for quoted aliases, so
  SELECT 1 AS "MyAlias", myalias is no longer resolved case-insensitively

See review comments:
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3281394375
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3281395049
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3281395713
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3281715932
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3281722319

yariks5s committed:
Address review: alias quote propagation, table-alias case, subcolumn per-part

- IQueryTreeNode::clone now copies alias_is_double_quoted, and IQueryTreeNode::toAST
  writes it back onto ASTWithAlias so clone / rebuild paths keep the flag and
  QueryExpressionsAliasVisitor does not silently register a quoted alias as
  case-insensitive after a clone
- alias_identifier_lookup uses the alias expression's own quote styles instead of
  the reference's, so SELECT "Col" AS x FROM t ORDER BY x resolves the cloned
  "Col" as case-sensitive even when x is unquoted
- Table-alias qualifier matching consults isAliasDoubleQuoted so that a
  double-quoted alias definition (FROM t AS "A") stays case-sensitive, even when
  the reference is unquoted. Applied in both tryBindIdentifierToTableExpression
  and tryResolveIdentifierFromTableExpression; the flag is propagated through
  QueryTreeBuilder for tables, subqueries, and table functions
- Subcolumn base lookup in tryResolveIdentifierFromStorage now keys off the
  first-part quote (the base column), so data."Name" matches base column Data
  case-insensitively while the tuple field Name stays exact-match
- Style fixes from CI: replaced "Db" example in ASTIdentifier comment and
  switched assert to chassert in tryResolveDatabaseNameCaseInsensitive

See review comments:
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3325978283
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3325978290
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3325978292
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3325978298

yariks5s committed:
Address review: per-part prefix quote, case-insensitive temporary table resolution

- tryResolveIdentifierFromExpressionArguments now splits the case-insensitivity
  flag: full-name lookup keeps last-part quoting, prefix lookup uses part-0
  quoting. Fixes arrayMap(item -> ITEM."Name", arr) where ITEM should still
  bind to lambda argument item even though "Name" is quoted

- Context gains tryResolveExternalTableNameCaseInsensitive for temporary/
  external tables; tryResolveTableIdentifier consults it before falling back to
  the current database. Now CREATE TEMPORARY TABLE Temp ...; SELECT * FROM temp
  resolves in standard mode

See review comments:
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3326569484
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3326569485

yariks5s committed:
Fix CI: track CTE name quote style, move settings-history entry to 26.6

Fast test failures on 99b9a45:
- 02995_new_settings_history expected case_insensitive_names to be registered
  under the current 26.6 version. Moved the history entry from 26.5 to 26.6
- 03774_case_insensitive_names_advanced expected case-insensitive CTE name
  collision to error at registration. My previous change kept the original
  case for all CTE names, which silently allowed WITH MyCTE AS (...), mycte
  AS (...). The bot's earlier comment (#3281394375) wanted quoted CTEs to
  stay case-sensitive, so I now thread the quote-style flag end-to-end and
  only fold unquoted CTEs into the lowercase index:
    * ASTWithElement.name_is_double_quoted populated from ParserWithElement
    * Propagated via CommonTableExpressionData into QueryNode/UnionNode
      (cte_name_is_double_quoted), and copied by cloneImpl
    * Registration in standard mode adds the CTE to
      lowercase_cte_to_original_names only when unquoted, and now throws
      MULTIPLE_EXPRESSIONS_FOR_ALIAS at registration when an unquoted CTE
      would collide with an existing one case-insensitively. Quoted CTEs
      remain case-sensitive distinct entries.

Report: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=92323&sha=99b9a45792625eaf0fc9482353980f5f9d03d5f7&name_0=PR&name_1=Fast%20test

yariks5s committed:
Refactor: consolidate case-insensitive identifier helpers and contracts

Clean-up pass following the iterative review. Behavior is unchanged.

- IdentifierLookup gains `isPartCaseInsensitive(part, standard_mode)` and
  `isLastPartCaseInsensitive(standard_mode)`. The 8+ ad-hoc copies of
  `standard_mode && !isPartDoubleQuoted(...)` across the analyzer now go
  through these helpers, putting the rule in one place
- ScopeAliases gains `registerAlias(context, name, node, register_for_ci_lookup)`
  which inserts into the primary alias map and (only on success) into the
  lowercase index. QueryExpressionsAliasVisitor, TableExpressionsAliasVisitor,
  and `add_table_expression_alias_into_scope` all go through it, so callers
  cannot write to one map and forget the other
- tryBindIdentifierToAliases and tryBindIdentifierToArrayJoinExpressions now
  key off part-0 quoting (the alias's own quote style), matching the fix
  applied to tryResolveIdentifierFromAliases in an earlier round
- tryResolveIdentifierFromCTE: registration enforces case-insensitive
  uniqueness for unquoted CTEs, so the lowercase index holds at most one
  original per key. The previously-unreachable ambiguity branch is replaced
  by a chassert
- Documentation: each `lowercase_*` parallel map now states what method must
  be used to insert/update it and what invariant the sibling primary map
  carries, so future code touching identifier handling has a hard-to-miss
  reminder
- tryResolveTableIdentifierFromDatabaseCatalog setup collapsed from a
  parts-size switch into a two-line expression

yariks5s committed:
Refactor: cache `standard_mode` query and encapsulate CTE registration

Reduces call-site noise from the iterative review:

- IdentifierResolveScope::isStandardMode() reads the setting once at the
  scope boundary; 17 sites that previously did
  `scope.context->getSettingsRef()[Setting::case_insensitive_names] == ::Standard`
  now read `scope.isStandardMode()` instead. Setting access is still on
  demand (no caching needed - the read is cheap), but the expression lives
  in one place

- IdentifierResolveScope::registerCTE(name, node, is_double_quoted) is the
  single entry point for adding a CTE to the scope. It updates both
  `cte_name_to_query_node` and (for unquoted CTEs in standard mode)
  `lowercase_cte_to_original_names`, and returns an enum signalling whether
  the name was new, duplicate, or a case-insensitive collision. The two-map
  insert+collision-check logic moves out of resolveQuery

yariks5s committed:
Fix temp/regular table precedence in standard-mode resolution; add test

Spotted in self-review: when an unqualified table name had both a temporary
and a regular table that case-collided, the regular table shadowed the
temporary one (default mode resolves temp first, so this was a behavior
inversion only in `standard` mode). Skip the current-database lookup once a
temporary-table match has been found.

Added a focused test exercising standard-mode temporary-table lookup with
unquoted/lowercase references and the double-quoted exact-case rule.

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/case-insensitive-identifiers

# Conflicts:
#	src/Analyzer/Resolve/IdentifierResolveScope.h
#	src/Analyzer/Resolve/IdentifierResolver.cpp
#	src/Analyzer/Resolve/QueryAnalyzer.cpp
#	src/Analyzer/Resolve/TableExpressionData.h
#	src/Parsers/ExpressionElementParsers.cpp

yariks5s committed:
Fix CI: track CTE name quote style, move settings-history entry to 26.6

CI on `76b6cd69d66`:

- ~14 stateless tests using `SETTINGS compatibility = '<old version>'` failed
  because `SettingsChangesHistory.cpp` carried two entries
  (`query_plan_min_columns_for_join_lazy_indexing`,
  `query_plan_max_limit_for_join_lazy_indexing`) that master reverted but
  the merge brought back. Settings-snapshot lookups for old compatibility
  values then tried to set these non-existent settings and errored. Removed
  the two stale entries; my `case_insensitive_names` entry stays
- `arm_tidy` build failed `cppcoreguidelines-init-variables`: the new
  `IdentifierQuoteStyle quote_style` variable in
  `ParserCompoundIdentifier::parseImpl` was conditionally assigned. Default
  to `None`

Bot review comments addressed:

- `ASTIdentifier::formatImplWithoutAlias` now honours per-part `quote_styles`
  on output, so a format/reparse round-trip preserves whether a part was
  double-quoted (and therefore case-sensitive in `standard` mode). Previously
  `"MyColumn"` could format back as unquoted `MyColumn`
- `IdentifierLookup` equality/hash now include `is_part_double_quoted`, so
  the identifier-resolve cache cannot serve a case-insensitive (unquoted)
  result to a later double-quoted lookup of the same text
- `canBindIdentifier` in both `tryBindIdentifierToTableExpression` and
  `tryResolveIdentifierFromTableExpression` now passes `part_case_insensitive(0)`
  instead of `column_case_insensitive`. The bind check keys off the base
  column part, not the trailing subcolumn, so `Data."Name"` can recognise
  unquoted `Data` as a possible `Data` column while the quoted `"Name"` field
  still gets exact-matched in the subcolumn lookup

See review comments:
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3429773741
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3429781304
  https://github.com/ClickHouse/ClickHouse/pull/92323#discussion_r3429791397

yariks5s committed:
Address review: qualified matchers, COLUMNS(...), alias/CTE format, table-expr lookup, doc scope

Findings from PR review:

1. `tryResolveIdentifierFromTableExpression` early `isTableExpressionLookup`
   branch was an exact-string compare. Qualified-matcher table resolution
   disables the database-catalog fallback, so an unquoted `testtable.*`
   could not find `TestTable` in standard mode. Now uses per-part case
   sensitivity from `IdentifierLookup`

2. Qualified matchers (`T.*`, `T.COLUMNS(...)`, `T.COLUMNS('regex')`)
   dropped qualifier quote styles when built. `MatcherNode` now carries
   `qualified_identifier_quote_styles` (and per-argument `columns_identifiers_quote_styles`);
   `QueryTreeBuilder` copies them from the AST; `resolveQualifiedMatcher`
   seeds `IdentifierLookup::is_part_double_quoted` from them, so `"T".*`
   stays case-sensitive in standard mode

3. `COLUMNS(identifier, ...)` was exact-case only. `MatcherNode::isMatchingColumn`
   now accepts a `standard_mode` flag; unquoted COLUMNS arguments fall back
   to a lowercase-set lookup, quoted arguments stay exact. All three call
   sites pass `scope.isStandardMode()`

4. Formatting of aliases and CTE names did not preserve their double quotes.
   `ASTWithAlias::formatImpl` and `ASTWithElement::formatImpl` now write
   the alias/CTE name with `writeDoubleQuotedString` when the original
   identifier was double-quoted, so format/reparse round-trips keep the
   semantic case-sensitivity in standard mode

5. `case_insensitive_names` docstring narrowed to the supported surface
   (SELECT-time analyzer resolution: db/table/column/alias/CTE). `INSERT`
   target/column-list lookup and DDL (`CREATE`/`DROP`/`ALTER`/`RENAME`)
   still go through exact `Context::resolveStorageID` + `Block::findByName`
   + catalog lookup. Documented as a known scope limitation; follow-up

Added test `100003_03774_case_insensitive_names_matchers_and_format` covering:
COLUMNS(unquoted) / COLUMNS("Quoted"), unquoted `testtable.*` against
`TestTable`, quoted `"TESTTABLE".*`, quoted alias and CTE round-trip survival,
and quoted `"MyCte"` / `"mycte"` not colliding.

yariks5s committed:
Rename test to fit numeric sequence; preserve only DoubleQuote in formatter

- Renames  to  so the test-numbering style check passes (gap <= 100)
- ASTIdentifier::formatImplWithoutAlias preserves DoubleQuote (semantically meaningful
  in standard mode) but lets Backtick fall through to the formatter's default quoting
  policy, so existing show-create / format tests that rely on backtick normalization
  do not regress

yariks5s committed:
Conservative format fix: preserve only DoubleQuote, let Backtick fall through

The earlier change preserved Backtick too, which produced spurious backticks
in the format output for plain identifiers and broke several show-create /
format-stability tests. Backticks behave like unquoted in all modes, so only
DoubleQuote needs to round-trip through format/reparse to keep case-sensitivity
in standard mode

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/case-insensitive-identifiers

# Conflicts:
#	src/Analyzer/Resolve/IdentifierResolver.cpp

yariks5s committed:
Update tests: keep double-quotes in format output, fix own-test row-count assertion

- 02868_select_support_from_keywords.reference: the CREATE VIEW used double-quoted
  identifiers (, ) for columns whose names are SQL keywords. The
  formatter now preserves the user's double-quoting (semantically required for
  case-sensitivity in standard mode) instead of normalizing to backticks
- 04356_03774_case_insensitive_names_matchers_and_format: my own test asserted
  count() = 0 for a COLUMNS("firstname") expansion against a table with no
  matching column, but count() returns the row count regardless of matched
  columns. Replaced with a sibling literal marker and LIMIT 1 so the assertion
  actually exercises whether the COLUMNS expansion matched anything

yariks5s committed:
Address review: COLUMNS list quote propagation, qualifier-binds case sensitivity, doc nit

- resolveUnqualifiedMatcher (COLUMNS_LIST branch): build the per-argument
  IdentifierLookup with `is_part_double_quoted` copied from the matcher's
  recorded `columns_identifiers_quote_styles`. Previously a quoted
  `COLUMNS("FirstName")` was resolved with no quote info, so in standard
  mode it could match a column `firstname` case-insensitively even though
  the user explicitly double-quoted

- qualifierBindsToJoinSubtree (used by
  analyzer_compatibility_prefer_alias_over_subcolumn): now per-call
  case-insensitive. Unquoted `B.id` against right-side alias `b` binds in
  standard mode; a double-quoted alias definition stays case-sensitive

- Settings.cpp: fix Markdown nesting in the `case_insensitive_names`
  example (the inner backticks around `FirstName` broke the surrounding
  code span; wrapped in a double-backtick span)

yariks5s committed:
Address review: Array(JSON) parser, Nested prefix, projection-override quotes

Closes the three open bot comments from this review round.

- ParserCompoundIdentifier: when the parser splices a synthetic `Array(JSON)`
  part into `name_parts`, also append `IdentifierQuoteStyle::None` to
  `quote_styles` so the per-part vector stays aligned. Otherwise the
  `DoubleQuote` style for a quoted tail like `."Mixed"` shifted onto the
  synthetic part

- tryResolveIdentifierAsNestedPrefix: prefix-part comparison now honours
  case-sensitivity. Passed `subcolumn_base_case_insensitive` from the
  caller so `Items Nested(...)` resolves an unquoted `items` reference in
  standard mode

- Projection-override aliases (`FROM (...) AS t("MyCol")`, `WITH cte("MyCol")`)
  now carry per-alias quote info:
    * QueryNode gains `projection_aliases_to_override_is_double_quoted`
    * QueryTreeBuilder records `IdentifierQuoteStyle::DoubleQuote` per alias
      from the AST and feeds both setters
    * AnalysisTableExpressionData::enableStandardMode accepts an exclusion
      set of column names to keep case-sensitive; the call site populates
      it from QueryNode's per-alias quote flags. Quoted overrides are
      excluded from `lowercase_column_name_to_original_names`, so unquoted
      `mycol` no longer matches a quoted `"MyCol"` definition

Test: COLUMNS("firstname") against no matching column now correctly errors
with UNKNOWN_IDENTIFIER (was silently empty). Updated own-test assertion.

yariks5s committed:
Address review: hash/equality include quote flags; per-part COLUMNS / column lookup; WITH/UNION quote propagation

Closes the 8 bot comments from this round; all are real semantic correctness
issues for `standard` mode that would otherwise let query-tree dedup/cache
collapse two queries that resolve differently.

Tree hash + equality include quote flags
- IQueryTreeNode::isEqual / getTreeHash include alias_is_double_quoted.
  `expr AS x` and `expr AS "x"` are now distinct nodes
- QueryNode::isEqualImpl / updateTreeHashImpl include
  cte_name_is_double_quoted, projection_aliases_to_override, and
  projection_aliases_to_override_is_double_quoted
- UnionNode::isEqualImpl / updateTreeHashImpl include
  cte_name_is_double_quoted
- MatcherNode::isEqualImpl / updateTreeHashImpl include
  qualified_identifier_quote_styles and columns_identifiers_quote_styles
- ASTIdentifier::updateTreeHashImpl includes quote_styles, so
  QueryResultCache no longer shares an entry between `"X"` and `X`

Per-part quote handling
- MatcherNode::isMatchingColumn compares the matcher identifier and the
  candidate column part-by-part: unquoted parts match case-insensitively,
  double-quoted parts must match exactly. `COLUMNS(data."Name")` matches
  `Data.Name` correctly
- tryResolveIdentifierFromStorage: when any qualifier part of a compound
  identifier is double-quoted, disable case-insensitive full-name lookup
  so `"data".Name` does not match `Data.Name`

Projection-override quote propagation
- WITH cte("MyCol") AS (...): collect double-quote flags alongside names
  in buildSelectExpression and call setProjectionAliasesToOverrideIsDoubleQuoted
- AnalysisTableExpressionData population: when the table expression is a
  UnionNode whose first inner QueryNode carries the override, drill into
  it to collect the case-sensitive (quoted) override names

Test
- 04064_format_cte_column_aliases.reference: the single-space alias `" "`
  now formats back as `" "` (preserved double-quote) instead of the older
  backtick normalization, matching the formatter contract from the
  previous round

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/case-insensitive-identifiers

yariks5s committed:
Address review: restoreTable + toAST quote propagation; drop no-parallel; preserve old hashes when quote flags are default

- ASTIdentifier::setShortName / restoreTable keep quote_styles aligned with name_parts when
  the helpers mutate name_parts (prepending the table qualifier or shrinking to a short name),
  so a quoted identifier does not get its DoubleQuote shifted onto an unintended part
- QueryNode::toASTImpl writes alias_is_double_quoted on the projection alias ASTWithAlias, so
  format/reparse of FROM (SELECT 1) AS t("MyCol") or WITH cte("MyCol") AS … emits "MyCol"
  and keeps the alias case-sensitive on reparse under standard mode
- Hash backward-compat: my earlier changes to QueryNode, UnionNode, MatcherNode,
  ASTIdentifier and IQueryTreeNode hash were unconditional and broke hash for nodes without
  any quote info — which the optimizer uses to order filter conditions
  (01625_constraints_index_append). The new material is now mixed in only when the
  corresponding flag is non-default, so old queries (no quoted aliases/CTE/COLUMNS/qualifier)
  keep the prior hash and filter order
- 03774 / 04356 tests: drop the unnecessary no-parallel tag from tests that touch only
  session settings and ordinary objects

yariks5s committed:
Address review: hash quote flags on ASTWithAlias/ASTWithElement; propagate quote styles in MatcherNode and CTE toAST

yariks5s committed:
Address review: case-aware alias self-ref, JOIN USING / ARRAY JOIN / lambda / recursive-CTE quote handling, subcolumn first-part case-sensitivity, info_schema alias dedup, external-table chain lookup, ASTIdentifier quote alignment

yariks5s committed:
Address review: exact-case-first lookup for aliases/columns/tables, lambda quote-style propagation, info_schema dedup; add 04365 regression test

yariks5s committed:
Address review: exact-case-first temp-table lookup, case-insensitive subcolumn suffix fold; extend 04365

yariks5s committed:
Address review: separate lowercase first-part index from case-sensitive set

yariks5s committed:
Address review: shadow-correct context walk in temp-table CI lookup; case-aware JOIN USING column-name comparison

yariks5s committed:
Address review: document standard-mode exact-case-first vs ambiguity in case_insensitive_names

yariks5s committed:
Address review: compact projection-override vectors on prune; suffix-quote-aware subcolumn fold; joinGet uses resolved storage name

yariks5s committed:
Address review: column transformer case-fold, INTERPOLATE/table-function/CTE-alias quote tracking, hasFullIdentifierName per-part flag, IDatabase exact-first table lookup

yariks5s committed:
Address review: hash quote_styles on IdentifierNode; round-trip quote for INTERPOLATE/REPLACE; AMBIGUOUS_IDENTIFIER on case-fold REPLACE collisions

yariks5s committed:
Address review: hash CTE alias children + INTERPOLATE target quote bit; fix joinGet regression vs temporary tables

yariks5s committed:
Fix CI 03520: IdentifierNode hash mixes only DoubleQuote, not Backtick (matches AST round-trip stability)

yariks5s committed:
Address review: MatcherNode hash only DoubleQuote; JOIN USING qualifier/alias quote awareness + duplicate detection; EXCEPT ambiguity on case-fold; QueryResultCache removeTableAliases preserves quote bits; information_schema canonicalize mixed-case views

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/case-insensitive-identifiers

jkartseva commented:
Is this PR ready for another look?

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/case-insensitive-identifiers

yariks5s committed:
Address review: joinGet skips canonical-name substitution for temporary tables

yariks5s committed:
Fix CI 04329 crash: restore qualified-matcher explicit-identifier type adjustment lost during merge

yariks5s committed:
Fix CI: restore correlated MATERIALIZED CTE check on identifier-resolution path; move case_insensitive_names to 26.7 settings history; ArrayJoinNode preserves alias quote bit; clang-tidy modernize-loop-convert

yariks5s committed:
Drop misleading comment: 04322 check restores master 55127470e8a lost in merge ef2de57c6f5

yariks5s committed:
Address review + fix CI crash: ASTIdentifier UAF, exact-first lambda lookup, ARRAY JOIN/EXCEPT quote awareness, remote-db skip, drop redundant hash/lowercase-set

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/case-insensitive-identifiers

# Conflicts:
#	src/Analyzer/Resolve/ScopeAliases.h

yariks5s committed:
Merge branch 'master' into yarik/case-insensitive-identifiers

yariks5s committed:
Address review: EXCEPT compound round-trip via multi-part identifier, REPLACE per-part quotes, document remote-db CI exclusion

yariks5s committed:
Merge remote-tracking branch 'origin/yarik/case-insensitive-identifiers' into yarik/case-insensitive-identifiers

# Conflicts:
#	src/Analyzer/Resolve/ScopeAliases.h

yariks5s committed:
Merge branch 'master' into yarik/case-insensitive-identifiers

yariks5s committed:
Address review: folded subcolumn canonical name, JOIN USING resolved-identity, INTERPOLATE/joinGet/remote-db quote fixes, MaterializedCTE name_is_double_quoted

yariks5s committed:
Merge remote-tracking branch 'origin/yarik/case-insensitive-identifiers' into yarik/case-insensitive-identifiers

yariks5s committed:
Address review: temp-vs-regular precedence, CTE quote on AnalysisTableExpressionData, REPLACE WHERE/HAVING folded rewrite, removeAlias clears quote, isLastPartDoubleQuoted via index, segfault wording, whitespace

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/case-insensitive-identifiers

# Conflicts:
#	src/Analyzer/Resolve/QueryAnalyzer.cpp

yariks5s committed:
Skip 03774_case_insensitive_names_db_table on darwin: case-different sibling DBs/tables collide on case-insensitive FS

yariks5s committed:
Merge branch 'master' into yarik/case-insensitive-identifiers

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/case-insensitive-identifiers

yariks5s committed:
Review round: structured EXCEPT/REPLACE target parts (fixes fuzzer !part.empty()), alias quote pins on all table-expression rewrites, quoted projection aliases, INTERPOLATE canonical target, USING resolved-identity, perf: lazy error strings, cached standard_mode, no per-identifier quote vector

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/case-insensitive-identifiers

# Conflicts:
#	src/Interpreters/DatabaseCatalog.h
#	tests/queries/0_stateless/02868_select_support_from_keywords.reference

yariks5s committed:
Fix flaky-check collision: move RENAME DATABASE case-fold test to .sh with unique per-instance names

yariks5s committed:
Simplify: IdentifierLookup::setQuoteFlagsFrom collapses 10 conversion loops (canonical empty=unquoted unifies resolve-cache keys), shared identifierPartsEqual, SubcolumnMatchFlags struct, extract projection-override helpers, IdentifierQuoteStyle to fwd header

yariks5s committed:
Fix 04499: collapse repeated UNKNOWN_DATABASE lines (client error verbosity differs per environment)

yariks5s committed:
Include projection alias quote flags in query node equality/hash/AST round-trip; honor CTE quote pin in table-expression shortcut

yariks5s committed:
Pin quoted recursive CTE columns on the self-reference table; simplification sweep across resolver, transformers, and catalog

yariks5s committed:
Key REPLACE transformer mappings by canonical column name only so quoted references cannot match folded target spellings

yariks5s committed:
Document CTE definition-time case-collision rule; note qualified REPLACE references are unchanged in both modes

yariks5s committed:
Key the identifier resolve cache by identifier parts so equal text with different part boundaries stays distinct

yariks5s committed:
Absorb canonical source keys when USING spelling is folded; pick one canonical Nested prefix with exact-first and ambiguity on case-siblings

yariks5s committed:
Resolve EXCEPT/REPLACE targets against the column namespace up front; fold NATURAL JOIN common-column intersection with exact-first

yariks5s committed:
Round 6: AST hash part boundaries, folded USING merge with participant identity, pre-resolution alias quote capture, NATURAL JOIN pins and sibling ambiguity, transformer parts sync

yariks5s committed:
Fix arm_tidy performance-move-const-arg; carry recursive CTE quote pins into RecursiveCTETable for AST round-trip

yariks5s committed:
Reject transformer targets that collapse onto one column after case folding

yariks5s committed:
Round 8: namespace-aware REPLACE filter rewrite, qualified COLUMNS exact-first, USING merge and per-side star suppression, recursive NATURAL pins, alias parser quote bits, DataLake guard, quote_styles lockstep

yariks5s committed:
Round 9: cross-scope REPLACE screening, sticky pins through pass-through projections, exact-first across join and cross-join sides, folded USING type adjustment, system-table cache policy with folded db names, TableNode quote pins, presentation quoting never pins

yariks5s committed:
Round 10: qualifier-aware exact-first, N-participant USING match, per-part transformer and subcolumn fold masks, materialized CTE pins on the deferred finalization path, presentation quoting diverts to backticks

yariks5s committed:
Add IdentifierName, FoldedNameIndex, and the database_and_table_name_matching setting

yariks5s committed:
Resolve database and table names through FoldedNameIndex in the catalog under standard matching

yariks5s committed:
Carry per-part identifier quoting from the parser through ASTIdentifier and StorageID into name resolution

yariks5s committed:
Add stateless test for standard database and table name matching

yariks5s committed:
Preserve quote pins on all StorageID paths, check access on canonical names, fold database-only statements, gate internal references

yariks5s committed:
Fold names in CREATE TABLE database component, EXISTS, SHOW COLUMNS and SHOW INDEXES under standard matching

yariks5s committed:
Write canonical table name back into ALTER before access collection

yariks5s committed:
Canonicalize before ON CLUSTER dispatch and access checks in EXCHANGE/RENAME/ALTER DATABASE/SYSTEM; pin implicit current database

yariks5s committed:
Fold CREATE AS source, ALTER partition secondary tables and view targets under standard matching

yariks5s committed:
Pin implicit database in EXISTS/SHOW COLUMNS/SHOW INDEXES; replay DDL log entries with exact matching

yariks5s committed:
Canonicalize names in legacy analyzer qualifier matching, SYSTEM dispatch, and CREATE/DROP INDEX rewrite

yariks5s committed:
Canonicalize unqualified DROP before cluster dispatch, fix UUID-race fallback spelling, skip exact-only overlay children

yariks5s committed:
Clarify that pinning is the predefined-alias mechanism, not a creation-time contract

yariks5s committed:
Document why quote styles stay out of the AST hash

yariks5s committed:
Reject unquoted case-sibling database and table creation under standard matching

yariks5s committed:
Serialize sibling creation on folded-name DDL guards and validate before cluster dispatch

yariks5s committed:
Enforce the case-sibling policy at all dispatch sites, after access checks

yariks5s committed:
Cover the old-format ON CLUSTER create path; document the best-effort scope of dispatch-side checks

yariks5s committed:
Add column_and_query_name_matching setting with quote plumbing into the query tree and analyzer guards

yariks5s committed:
Fold column and subcolumn references in the analyzer under standard matching

yariks5s committed:
Fold aliases, CTE names and lambda arguments under standard matching with definition-collision rejection

yariks5s committed:
Fold JOIN USING, NATURAL JOIN, matcher qualifiers and column transformers under standard matching

yariks5s committed:
Canonicalize table references in stored view bodies, reject folded column spellings; add grants and persistence tests

yariks5s committed:
Supersede the exact-first prototype with the standard-matching foundation

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/identifier-matching-foundation

# Conflicts:
#	src/Analyzer/Resolve/QueryAnalyzer.cpp

yariks5s committed:
Thread matching mode into the constant-expression alias visitor from master merge

yariks5s committed:
Fix style and tidy findings: initialize quote variables, deduplicate error code, annotate tolerated catch, doc wording

yariks5s committed:
Pin quoted projection aliases in subqueries and recursive CTEs; fold INTERPOLATE targets and prune by canonical name

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/identifier-matching-foundation

yariks5s committed:
Wrap fuzzer identifier parts from the master merge in IdentifierPart

yariks5s committed:
Fix logical error on unqualified CREATE TABLE ON CLUSTER: fill the current database before the case-sibling check

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/identifier-matching-foundation

yariks5s committed:
Split identifier-matching tests to fit the flaky-check time budget

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/identifier-matching-foundation

yariks5s committed:
Pin quoted CTE and temporary table names in qualifier lookup; close the query-tree hash and equality gap

yariks5s committed:
Renumber identifier-matching tests to avoid prefix collisions with master

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/identifier-matching-foundation

# Conflicts:
#	src/Common/QueryFuzzer.cpp

yariks5s committed:
Fold ARRAY JOIN aliases, canonicalize INTERPOLATE targets on the deferred projection path, close alias quote identity gap

yariks5s committed:
Keep default query-tree hashes master-identical; reapply transformer target quotes in AST conversion

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/identifier-matching-foundation

yariks5s committed:
Reject REPLACE transformer targets folding to the same column

yariks5s committed:
Acquire folded DDL guard after the exact-name guard to keep sorted lock order

yariks5s committed:
Fix tests for CI: no-fasttest tags, force analyzer, deterministic order (https://github.com/ClickHouse/ClickHouse/pull/92323)

yariks5s committed:
Merge remote-tracking branch 'origin/master' into yarik/identifier-matching-foundation

# Conflicts:
#	src/Analyzer/Resolve/QueryAnalyzer.cpp

yariks5s committed:
Keep the temporary table name quote in TableNode::toASTIdentifier

yariks5s committed:
Shorten and prune comments across the diff

yariks5s committed:
Restore Ok token in catch-all comment for the style check
